### PR TITLE
Disallow escape sequences in contextual keywords

### PR DESCRIPTION
**Pull request by danez**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          | y
| Major: Breaking Change?  | n
| Minor: New Feature?      | n
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

Resolves two problems:
- Disallow (async)(a) => {}
- Disallow escape sequences in contextual keywords like `async, as, of, get, set, static, ...`

The cases which were parsing successfully with escape sequences (although shouldn't) were: async in almost all cases. get & set in classes.

Also added tests for a lot of cases that were already correctly working.

### CODE REVIEW

File: packages/babel-parser/src/parser/expression.js [hunk 1]
```diff
@@ -704,11 +712,11 @@ export default class ExpressionParser extends LValParser {
 
   atPossibleAsync(base: N.Expression): boolean {
     return (
-      !this.state.containsEsc &&
```
Comment (danez) [inline packages/babel-parser/src/parser/expression.js]: `containsEsc` only says if the current token has escape sequences. This references the token after `async` which is the wrong one.

File: packages/babel-parser/src/parser/expression.js [hunk 2]
```diff
@@ -576,13 +586,8 @@ export default class ExpressionParser extends LValParser {
         this.expect(tt.bracketR);
         return this.finishNode(node, "OptionalMemberExpression");
       } else if (this.eat(tt.parenL)) {
-        const possibleAsync = this.atPossibleAsync(base);
-
         node.callee = base;
-        node.arguments = this.parseCallExpressionArguments(
-          tt.parenR,
-          possibleAsync,
-        );
```
Comment (danez) [inline packages/babel-parser/src/parser/expression.js]: I removed this, as I don't see a way what an `optionalCallExpression` has anything todo with the `async` keyword. No tests failed or changed. I guess this was copied wrong.

File: packages/babel-parser/src/parser/expression.js [hunk 3]
```diff
@@ -525,12 +525,21 @@ export default class ExpressionParser extends LValParser {
     startLoc: Position,
     noCalls?: ?boolean,
   ): N.Expression {
+    const maybeAsyncArrow = this.atPossibleAsync(base);
```
Comment (danez) [inline packages/babel-parser/src/parser/expression.js]: moving it from line 624 to here, basically prevents `(async)(a) => {}`

File: packages/babel-parser/src/parser/expression.js [hunk 4]
```diff
@@ -641,7 +645,11 @@ export default class ExpressionParser extends LValParser {
         this.finishOptionalCallExpression(node);
       }
 
-      if (possibleAsync && this.shouldParseAsyncArrow()) {
+      if (
+        maybeAsyncArrow &&
+        !this.canInsertSemicolon() &&
```
Comment (danez) [inline packages/babel-parser/src/parser/expression.js]: `!this.canInsertSemicolon()` prevents 
```js
async (x) 
=> {}
```

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/parser/expression.js]: Maybe we could move the check inside `shouldParseAsyncArrow`?
Also note that both flow and TS allow this code:
```js
async (f)
: t => { }
```
but disallow

```js
async (f) : t
=> { }
```

Comment (danez) [reply packages/babel-parser/src/parser/expression.js]: Thanks, I added testcases and moved it into `shouldParseAsyncArrow `

**Pull request merged**